packages/firestore/src/core/pipeline-util.ts

```diff
@@ -242,41 +288,322 @@ export function toPipeline(query: Query, db: Firestore): Pipeline {
+    let result = `${stage.name}(${canonifyExpr(stage._field)},${
+      stage._distanceMeasure
+    },[${vector}]`;
+    if (!!stage._limit) {
+      result = result + `,${stage._limit}`;
+    }
+    if (!!stage._distanceField) {
+      result = result + `,${stage._distanceField}`;
+    }
+    return result + ')';
+  }
+  if (stage instanceof Limit) {
+    return `${stage.name}(${stage.limit})`;
+  }
+  if (stage instanceof Offset) {
+    return `${stage.name}(${stage.offset})`;
+  }
+  if (stage instanceof Select) {
+    return `${stage.name}(${canonifyExprMap(stage.projections)})`;
+  }
+  if (stage instanceof Sort) {
+    return `${stage.name}(${canonifySortOrderings(stage.orders)})`;
+  }
+
+  throw new Error(`Unrecognized stage ${stage.name}`);
+}
+
+function canonifyExprMap(map: Map<string, Expr>): string {
+  const sortedEntries = Array.from(map.entries()).sort();
+  return `${sortedEntries
+    .map(([key, val]) => `${key}=${canonifyExpr(val)}`)
+    .join(',')}`;
+}
+
+export function canonifyPipeline(p: CorePipeline): string;
+export function canonifyPipeline(p: CorePipeline): string {
+  return p.stages.map(s => canonifyStage(s)).join('|');
+}
+
+// TODO(pipeline): do a proper implementation for eq.
+export function pipelineEq(left: CorePipeline, right: CorePipeline): boolean {
+  return canonifyPipeline(left) === canonifyPipeline(right);
+}
+
+export type PipelineFlavor = 'exact' | 'augmented' | 'keyless';
+
+export type PipelineSourceType =
+  | 'collection'
+  | 'collection_group'
+  | 'database'
+  | 'documents';
+
+export function asCollectionPipelineAtPath(
+  pipeline: CorePipeline,
+  path: ResourcePath
+): CorePipeline {
+  const newStages = pipeline.stages.map(s => {
+    if (s instanceof CollectionGroupSource) {
+      return new CollectionSource(path.canonicalString());
+    }
+
+    return s;
+  });
+
+  return new CorePipeline(pipeline.serializer, newStages);
+}
+
+export type QueryOrPipeline = Query | CorePipeline;
+
+export function isPipeline(q: QueryOrPipeline): q is CorePipeline {
+  return q instanceof CorePipeline;
+}
+
+export function stringifyQueryOrPipeline(q: QueryOrPipeline): string {
+  if (isPipeline(q)) {
+    return canonifyPipeline(q);
+  }
+
+  return stringifyQuery(q);
+}
+
+export function canonifyQueryOrPipeline(q: QueryOrPipeline): string {
+  if (isPipeline(q)) {
+    return canonifyPipeline(q);
+  }
+
+  return canonifyQuery(q);
+}
+
+export function queryOrPipelineEqual(
+  left: QueryOrPipeline,
+  right: QueryOrPipeline
+): boolean {
+  if (left instanceof CorePipeline && right instanceof CorePipeline) {
+    return pipelineEq(left, right);
+  }
+  if (
+    (left instanceof CorePipeline && !(right instanceof CorePipeline)) ||
+    (!(left instanceof CorePipeline) && right instanceof CorePipeline)
+  ) {
+    return false;
+  }
+
+  return queryEquals(left as Query, right as Query);
+}
+
+export type TargetOrPipeline = Target | CorePipeline;
```

Comment:
Question to be covered with the team. But what allows us to skip creating of an equivalent "Pipeline Target" class